cryptoki/src/types.rs

```diff
@@ -208,3 +252,110 @@ impl From<CK_VERSION> for Version {
         }
     }
 }
+
+/// A UTC datetime returned by a token's clock if present.
+#[derive(Copy, Clone, Debug)]
+pub struct UtcTime {
+    /// **[Conformance](crate#conformance-notes):**
+    /// Guaranteed to be in range 0..=9999
+    pub year: u16,
```

Comment:
:thinking: Not sure if it's correct to say it's guaranteed to be in range if it's `pub`. I guess what you mean is that the PKCS#11 lib will not return other values but I'm not sure how to approach this. Make this struct immutable and only expose fields via impl functions?

Reply:
Yes. I meant the original format for this type is a character string with exactly 4 digits for this field. If it doesn't land in that range, the conversion will fail and generate an error, never constructing the type at all. I'm definitely open to alternatives. Did you happen to see my [comment](5dfddf8) on this? Because I definitely considered several.

Reply:
Ha, no I haven't. Now that I've read them I understand the problem better. FTR I'd still use getters just in case the struct will be used as an input one day. We'd avoid the API breakage. One more thing would be additional `to_iso_string` function. I know Display does that but from the API discoverability perspective when someone wants the ISO time they'd be sure `to_iso_string` would do that while for Display they'd have to read the docs and still be unsure if the format won't change.

Still, cool work. I'll be watching this PR for further changes. 👍

Reply:
> FTR I'd still use getters just in case the struct will be used as an input one day. We'd avoid the API breakage.

I can do this, but how would you feel about making the fields pub and [read-only](https://docs.rs/readonly/latest/readonly/index.html) instead of getters? I'm concerned a bit about callback bloat in ancillary types. This one in particular doesn't seem to have changed or found new use in the 3.x drafts that will eventually come into play, and I don't think _any_ kind of change is being made to the 2.x standard any more.

> One more thing would be additional to_iso_string function. I know Display does that but from the API discoverability perspective when someone wants the ISO time they'd be sure to_iso_string would do that while for Display they'd have to read the docs and still be unsure if the format won't change.

Glad to hear another opinion on "How much ISO (if any) is appropriate?" I like the suggestion of having a named function instead. Would you recommend removing Display altogether then?

Reply:
> I can do this, but how would you feel about making the fields pub and read-only instead of getters?

This is 👌 too. 

> I'm concerned a bit about callback bloat in ancillary types.

Do you mean the overhead of function calls for simple getters? I guess they should be inlined anyway, either by default or with a [little bit of help](https://nnethercote.github.io/perf-book/inlining.html). 

> Glad to hear another opinion on "How much ISO (if any) is appropriate?" I like the suggestion of having a named function instead. Would you recommend removing Display altogether then?

I don't mind the Display but it seems the docs indicate it's for human consumption not any stable format. I see you've removed it. That's OK too. If someone needs it it can be added later.